src/Tools/BuildValidator/LocalReferenceResolver.cs

```diff
@@ -77,110 +77,88 @@ public IEnumerable<AssemblyInfo> GetCachedAssemblyInfos(string fileName) => _nam
-                using var fileStream = File.OpenRead(filePath);
+            using var fileStream = File.OpenRead(assemblyInfo.FilePath);
 
-                // This is deliberately using an ordinal comparison here. The name of the assembly is written out 
-                // into the PDB. Rebuild will only succeed if the provided reference has the same name with the
-                // same casing
-                if (Path.GetFileName(filePath) != reference.Name)
-                {
-                    filePath = Path.Combine(Path.GetDirectoryName(filePath)!, reference.Name);
-                }
-                builder.Add(MetadataReference.CreateFromStream(
-                    fileStream,
-                    filePath: filePath,
-                    properties: new MetadataReferenceProperties(
-                        kind: MetadataImageKind.Assembly,
-                        aliases: reference.ExternAlias is null ? ImmutableArray<string>.Empty : ImmutableArray.Create(reference.ExternAlias),
-                        embedInteropTypes: reference.EmbedInteropTypes)));
+            // This is deliberately using an ordinal comparison here. The name of the assembly is written out 
+            // into the PDB. Rebuild will only succeed if the provided reference has the same name with the
+            // same casing
+            var filePath = assemblyInfo.FilePath;
+            if (Path.GetFileName(filePath) != metadataReferenceInfo.FileName)
+            {
+                filePath = Path.Combine(Path.GetDirectoryName(filePath)!, metadataReferenceInfo.FileName);
             }
 
-            results = builder.MoveToImmutable();
+            metadataReference = MetadataReference.CreateFromStream(
+                fileStream,
+                filePath: filePath,
+                properties: new MetadataReferenceProperties(
+                    kind: MetadataImageKind.Assembly,
+                    aliases: metadataReferenceInfo.ExternAlias is null ? ImmutableArray<string>.Empty : ImmutableArray.Create(metadataReferenceInfo.ExternAlias),
+                    embedInteropTypes: metadataReferenceInfo.EmbedInteropTypes));
             return true;
         }
 
-        public bool CacheNames(ImmutableArray<MetadataReferenceInfo> references)
+        public bool TryGetAssemblyInfo(MetadataReferenceInfo metadataReferenceInfo, [NotNullWhen(true)] out AssemblyInfo? assemblyInfo)
         {
-            if (references.All(r => _mvidMap.ContainsKey(r.Mvid)))
+            if (_mvidMap.TryGetValue(metadataReferenceInfo.ModuleVersionId, out assemblyInfo))
             {
-                // All references have already been cached, no reason to look in the file system
                 return true;
             }
 
+            if (_nameMap.TryGetValue(metadataReferenceInfo.FileName, out var list))
+            {
+                // File name has already been searched for and it was not found so no more searching
```

Comment:
>File name has already been searched for and it was not found [](start = 19, length = 60)

I don't understand this comment. It sounds like `_nameMap.TryGetValue(metadataReferenceInfo.File, ...)` did find the file name since `TryGetValue(...)` returned `true`. Should the comment refer to the MVID instead of the file name?

Reply:
Cleaned up the comment. 

Essentially the `_nameMap` tracks all of the file names we have searched for. When we are populating the cache for a metadata reference, say `Util.dll`, we look at every single `Util.dll` in the search paths. Then we populate a MVID map based on the result. This comment is essentially saying the MVID lookup failed and we've already populated DLLs with this name hence there is no reason to do more work.